Fix ordering of VS_MapList map insertion

PhasecoreX opened the pull request:
Also fixes IgnoredMaps not working
Also skips IgnoredMaps checks if size is zero

PhasecoreX committed:
Fix ordering of VS_MapList map insertion

Also fixes IgnoredMaps not working
Also skips IgnoredMaps checks if size is zero

PhasecoreX commented:
There might be one other thing that needs to be done, as `ML.RemoveMap(Game.default.MapPrefix$"-Tutorial");` doesn't seem to work with `CTF-Tutorial`, not sure.

Deaod commented on Classes/VS_MapList.uc:
```diff
@@ -46,6 +46,8 @@ function AddMap(string MapName) {
 		return;
 
+	if (i < Maps.Length && Caps(Maps[i]) < Caps(MapName))
+		i += 1;
```
I do not understand this change. Why is it necessary?

PhasecoreX replied:
`FindIndexForMap()` returns zero every time when the map list is size 1, as there's only one entry (index 0). This check is to see if the insert should happen before or after that 0 entry, which is needed to keep the alphabetical order of the list.

I've moved the logic into `FindIndexForMap()` itself. Now it returns the exact index if there's a match, the found index if the searched for map is less than the map in the list (should be inserted before, the current logic), or index + 1 if it's greater (should be inserted after the found map).

Deaod replied:
Thanks for the explanation.

PhasecoreX committed:
Fix Tutorial map handling

Check map list length before removing maps
Move map length 1 logic into FindIndexForMap()

Deaod commented on Classes/MutVoteSys.uc:
```diff
@@ -1178,8 +1178,6 @@ function VS_MapList LoadMapList(class<GameInfo> Game, name ListName) {
 	// As before, see if the list already exists for the specified game type.
 	ML = LoadMapListByPrefix(Game.default.MapPrefix);
 
-	ML.RemoveMap(Game.default.MapPrefix$"-Tutorial");
```
I believe this line should stay because i dont want to remove every map that ends with "-Tutorial", just from MapLists created for game modes.

Deaod commented on Classes/MutVoteSys.uc:
```diff
@@ -1238,10 +1236,16 @@ function VS_MapList LoadMapListByPrefix(string Prefix) {
 
 	do {
-		ML.AddMap(Left(MapName, Len(MapName) - 4)); // we dont care about extension
+		ML.AddMap(CleanMapName(MapName));
```
While not wrong, we know for sure that the map will have .unr at the end. CleanMapName thus does an unnecessary check which will execute more bytecode instructions.

This is not a deal breaker to me. I just thought i should outline my reason for not doing this before.

PhasecoreX replied:
That's fair. Usually I reuse code as much as possible with other languages, so logic is in one place, but I will revert this.

Deaod commented on Classes/MutVoteSys.uc:
```diff
@@ -1238,10 +1236,16 @@ function VS_MapList LoadMapListByPrefix(string Prefix) {
+		ML.RemoveMap(Prefix$"Tutorial");
+	} else {
+		ML.RemoveMap(Prefix$"-Tutorial");
+	}
```
I'm not a fan of this. Theoretically you could have a map just called Tutorial, without it being designed as a tutorial level for a particular game mode.

Is this necessary for anything in particular?

PhasecoreX replied:
It seems like if you don't specify a map list, or you specify a map list with `IncludeMapsWithPrefix=AS`, it adds all maps except `AS-Tutorial`. But, if you specify it as `IncludeMapsWithPrefix=AS-`, it will include `AS-Tutorial`.

My use case was specifically CTF, where I wanted a map list with CTF but no CTFM. So, I specified `IncludeMapsWithPrefix=CTF-`.

I've reverted this for now so you can accept the pull request. Not sure if this "CTF" vs "CTF-" thing needs to be addressed, it's just a weird edge case. And for me at least, CTF-Tutorial isn't a bad map, so I manually include it anyway. But, you wouldn't want this happening with AS-Tutorial.

Deaod replied:
I think i see an issue: Since the unification of Prefix MapLists and GameMode MapLists, whenever you specify a preset without maplist first (so it loads the game modes list via prefix), it then removes the -Tutorial map. If you then have a preset with a maplist that loads the same prefix as the game mode, youll get a list without -Tutorial (which is not what I intended). If you do it the other way around you get a game modes MapList with the -Tutorial map still on it.

Im not sure about a fix for this. Ill have to meditate on it.

Deaod commented on Classes/VS_MapList.uc:
```diff
@@ -17,8 +17,8 @@ function int FindIndexForMap(string MapName) {
 		Index = F + ((C - F) / 2);
+		CompMap = Caps(Maps[Index]);
 		if (CompMap < MapName)
 			F = Index+1;
```
```suggestion
			F = ++Index;
```
This would make the change down below unnecessary.

PhasecoreX replied:
Tested, yes that works!

Deaod commented on Classes/VS_MapList.uc:
```diff
@@ -27,6 +27,9 @@ function int FindIndexForMap(string MapName) {
 	}
 
+	if (Index < Maps.Length && Caps(Maps[Index]) < Caps(MapName))
+		Index += 1;
```
This becomes unnecessary with the suggested change above.

PhasecoreX replied:
Will remove

Deaod reviewed (commented):
First of all, thank you for taking the time to write this.

I would strongly urge you to consider the suggested change to `FindIndexForMap`.

The other comments im not firm on. Im happy to accept the rest of the changes if you believe they are still the best choice.

Let me know what you want to do.

PhasecoreX committed:
Cleaner  FindIndexForMap() implementation

PhasecoreX committed:
Revert Tutorial map handling
